### PR TITLE
[DO NOT SQUASH] llvm update 4/12/2022

### PR DESCRIPTION
**Pull request by jerryyin**
Summary of key changes:
 - `std` dialect has been fully migrated into `func` dialect
   - Impacted ops become: `func.call`, `func.constant`, `func.func` and `func.return`
   - `stdToLLVM` conversion pass become `FuncToLLVM`
 - `verifier` implementation can no longer stay within tablegen, and has to be defined seperately
   - Instead, one needs to specify `hasVerifier = 1` to generate a declaration
 - Fairly significant changes in bufferization pipelines
   - Refer to changes [here ](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/blob/827644ed28c818f762585f5ca943bf606c79f959/external/llvm-project/mlir/test/Dialect/Tosa/tosa-partition-run.mlir)
 - Included back the `CMAKE_INCLUDE_CURRENT_DIR`
   - After the upstream has gotten rid of equivalent one in tablegen: [here](https://reviews.llvm.org/rG03078ec20b12605fd4dfd9fe9c98a26c9d2286d7).
 - A conversion pass targeted on `funcOp` cannot be run together with the ones targeted on `ModuleOp`
   - To run from pipeline, you need to attach `func.func` like `-pass-pipeline="func.func(tosa-to-linalg-named)"`
   - To add into pass pipeline implementation, you need to explicitly use a nested pass like: `pm.addNestedPass<mlir::FuncOp>(createTensorBufferizePass())`

### CODE REVIEW

**Comment by jerryyin** [comment]
Please review, this PR has passed all necessary CI test except a few infrastructure related failures:
 - nightly CI: http://rocmhead.amd.com:8080/blue/organizations/jenkins/MLIR%2Fmlir/detail/PR-597/3/pipeline/262
 - PR CI: http://rocmhead.amd.com:8080/blue/organizations/jenkins/MLIR%2Fmlir/detail/PR-597/2/pipeline

**Comment by whchung** [comment]
@jerryyin could you help summarize major changes needed to keep up with upstream in this iteration?

**Comment by sjw36** [comment]
The page times out trying to load for review..

File: mlir/test/fusion/tosa-to-miopen-relu.xmir.mlir [hunk 1]
```diff
@@ -1,4 +1,5 @@
 // RUN: mlir-miopen-driver -host-pipeline full -kernel-pipeline full -triple amdgcn-amd-amdhsa -target gfx908 %s | FileCheck %s
+// XFAIL: *
```
Comment (jerryyin) [inline mlir/test/fusion/tosa-to-miopen-relu.xmir.mlir:2]: FYI: Will address as a independent topic.

Upstream (https://reviews.llvm.org/D119444) break us by removing essential lowering patterns. Needs to find out the equivalent available ones.

**Comment by jerryyin** [comment]
@whchung For sure, let me give a summary and update it in PR description.

**Review by sjw36 — APPROVED** [review]
Looks good.

File: mlir/test/fusion/tosa-to-miopen-relu.xmir.mlir (hunk 1 above)
Comment (sjw36) [reply mlir/test/fusion/tosa-to-miopen-relu.xmir.mlir:2]: Please link the ticket here.

Comment (jerryyin) [reply mlir/test/fusion/tosa-to-miopen-relu.xmir.mlir:2]: https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/521

**Pull request merged**